langchain4j-oracle-database/src/main/java/dev/langchain4j/store/embedding/oracle/OracleEmbeddingStore.java

```diff
@@ -0,0 +1,589 @@
+ * </tr><tr>
+ *     <td>id</td>
+ *     <td>VARCHAR(36)</td>
+ *     <td>
+ *         Primary key. Used to store {@link UUID} strings which are generated by {@link #add(Embedding)},
+ *         {@link #add(Embedding, TextSegment)}, {@link #addAll(List)}, and {@link #add(Embedding, TextSegment)}
+ *     </td>
+ *     </tr><tr>
+ *     <td>embedding</td>
+ *     <td>VECTOR(*, FLOAT32)</td>
+ *     <td>
+ *         Stores the {@link Embedding#vector()} passed to {@link #add(Embedding)},
+ *         {@link #add(Embedding, TextSegment)}, {@link #addAll(List)}, and {@link #add(Embedding, TextSegment)}.
+ *         Never stores NULL.
+ *     </td>
+ *     </tr><tr>
+ *     <td>content</td>
+ *     <td>CLOB</td>
+ *     <td>
+ *         Stores the {@link TextSegment#text()} passed to {@link #add(Embedding, TextSegment)} and
+ *         {@link #addAll(List, List)}. Stores NULL when {@link #add(Embedding)} and {@link #addAll(List)}
+ *         are called.
+ *     </td>
+ *     </tr><tr>
+ *     <td>metadata</td>
+ *     <td>JSON</td>
+ *     <td>
+ *         Stores the {@link TextSegment#metadata()} passed to {@link #add(Embedding, TextSegment)} and
+ *         {@link #addAll(List, List)}. Stores NULL when {@link #add(Embedding)} and {@link #addAll(List)}
+ *         are called.
+ *     </td>
+ * </tr></table>
+ * </p><p>
+ * An inverted flat file (IVF) vector index is created on the embedding column. The index is named
+ * "{tableName}_vector_index", where {tableName} is the name configured using the {@link Builder}.
+ * </p><p>
+ * All methods of this embedding store will throw an {@link IllegalStateException} with a {@link SQLException} as its
+ * {@link IllegalStateException#getCause()} when a database operation results in an error.
+ * </p>
+ */
+public final class OracleEmbeddingStore implements EmbeddingStore<TextSegment> {
+
+    /**
+     * The mapping function for use with {@link SQLFilters#create(Filter, UnaryOperator)}. The function maps a
+     * {@link Metadata} key to a field of the JSON "metadata" column. The builtin JSON_VALUE function is used to
+     * evaluate a JSON path expression.
+     */
+    private static final UnaryOperator<String> METADATA_KEY_MAPPER = key -> "JSON_VALUE(metadata, '$." + key + "')";
+
+    /** DataSource configured to connect with an Oracle Database. */
+    private final DataSource dataSource;
+
+    /** Name of a database table accessed by this embedding store */
+    private final String tableName;
+
+    /**
+     * Constructs embedding store configured by a builder.
+     *
+     * @param builder Builder that configures the emebedding store. Not null.
+     *
+     * @throws IllegalArgumentException If the configuration is not valid.
+     */
+    private OracleEmbeddingStore(Builder builder) {
+        this.dataSource = ensureNotNull(builder.dataSource, "dataSource");
+        this.tableName = ensureNotNull(builder.tableName, "tableName");
+
+        createSchema(dataSource, tableName);
```

Comment:
We use an `IF EXISTS` clause in our `CREATE TABLE` SQL. This means we don't get an error if the table already exists.
Or is the `createTable` option something that _recreates_ the table (and index), if they already exist? Or is the idea to just avoid an extra SQL call on creation of the embedding store?

Reply:
> Or is the `createTable` option something that recreates the table (and index), if they already exist?

No, I meant "create if does not exists"

> Or is the idea to just avoid an extra SQL call on creation of the embedding store?

Not really.
It is definitely a good dev experience for `OracleEmbeddingStore` to create the table automatically on dev environment. But I am not sure users will be happy if this call will be done on prod environment (without an option to disable it) when DDL is most probably managed in other ways and database user does not have permission to create a new table.

WDYT?

Reply:
I really appreciate this comment. I think we definitely need to design for the case where tables and indexes are already created.

Reply:
I guess adding a `createTable` flag (with default value `false`) in the builder/ctor should be enough

Reply:
Definitely want the default to be no table or index creation.

I was thinking there could be at least 3 desirable behaviors:
```java
/** An option for how a schema object, such as a table, is created. */
enum CreateOption {

   /** No attempt is made to create the schema object. */
   NONE,

  /** An existing schema object is reused, otherwise it is created */
  IF_NOT_EXISTS,

  /** An existing schema object is dropped and replaced with a new one */
  CREATE_OR_REPLACE

}
```
With the enum, we don't need:
```java
OracleEmbeddingStore.builder()
  .dropExistingTable(true)
  .createTable(true)
  .dropExistingIndex(true)
  .createIndex(true)
  ...
```
Instead:
```java
OracleEmbeddingStore.builder()
  .tableCreation(CREATE_OR_REPLACE)
  .indexCreation(CREATE_OR_REPLACE)
  ...
```
The enum makes the code a bit more expressive and readable.
An enum also means we can add new options in the future, without cluttering-up the builder interface with new methods. 
Not sure how much of an issue it would be, but the enum does prevent invalid combinations of builder calls, such as:
```java
OracleEmbeddingStore.builder()
  .dropExistingTable(true)
  .createTable(false)
  ...
```

Reply:
Makes sense!

Reply:
We added a `CreateOption` enum. The API ended like this:
```java
OracleEmbeddingStore.builder()
  .embeddingTable("my_embedding_table", CreateOption.CREATE_IF_NOT_EXISTS)
  .vectorIndex(CreateOption.CREATE_OR_REPLACE)
```